control-plane/agents/src/bin/core/controller/reconciler/nexus/mod.rs

```diff
@@ -130,61 +136,139 @@ async fn nexus_reconciler(
-                if let Err(error) = nexus
-                    .remove_child_by_uri(
-                        context.registry(),
-                        &nexus_state,
-                        &child.uri,
-                        true,
-                    )
-                    .await
-                {
-                    nexus_spec_clone.error_span(|| {
-                        tracing::error!(
-                            error = error.full_string().as_str(),
-                            %child.state,
-                            %child.state_reason,
-                            child.uri = %child.uri.as_str(),
-                            "Failed to remove faulted child"
-                        )
-                    });
-                } else {
-                    nexus_spec_clone.info_span(|| {
-                        tracing::info!(
-                            %child.uri,
-                            %child.state,
-                            %child.state_reason,
-                            "Successfully removed faulted child",
-                        )
-                    });
-                }
+        for child in nexus_state.children.iter().filter(|c| c.state.faulted()) {
+            if child.state_reason != ChildStateReason::NoSpace {
+                let _ = handle_child_rebuild(nexus, child, &nexus_state, context).await;
             }
         }
-        .instrument(span)
+    }
+    Ok(PollerState::Idle)
+}
+
+/// Checks rebuild policy for the child. Depending on the policy performs the either waits for
+/// child to comeback for specified duration or does Full rebuild straightaway.
+async fn handle_child_rebuild(
+    nexus_spec: &mut OperationGuardArc<NexusSpec>,
+    child: &Child,
+    nexus: &Nexus,
+    context: &PollContext,
+) -> Result<(), SvcError> {
+    let wait_duration = RuleSet::faulted_child_wait(nexus, context.registry());
+    if wait_duration.is_zero() || child.state_reason == ChildStateReason::RebuildFailed {
+        info!(%child.uri, "Start full rebuild for child");
+        faulted_children_remover(nexus_spec, &child.uri, context).await?
+    } else if is_time_elapsed(child.faulted_at, wait_duration, &child.uri).await {
+        info!(%child.uri, "Start full rebuild for child,  Wait time has elapsed");
+        faulted_children_remover(nexus_spec, &child.uri, context).await?;
+    } else if get_child_replica(nexus_spec.as_ref(), child, context)
         .await
+        .is_err()
+    {
+        info!(%child.uri,"Replica is still not online for child");
+    } else {
+        info!(%child.uri, "Replica for child is Online, starting partial rebuild");
+        if let Err(error) = initialize_partial_rebuild(&nexus.uuid, &child.uri, context).await {
+            info!(%child.uri, "Initializing Full rebuild as partial rebuild failed for child, failed with : {:?}", error);
+            faulted_children_remover(nexus_spec, &child.uri, context).await?
+        }
     }
+    Ok(())
+}
```

Comment:
I think we can simplify this a bit:

```suggestion
async fn handle_faulted_child(
    nexus_spec: &mut OperationGuardArc<NexusSpec>,
    child: &Child,
    nexus: &Nexus,
    context: &PollContext,
) -> Result<(), SvcError> {
    // Child is permanently faulted, we have remove it and rebuild anew.
    if child.state_reason == ChildStateReason::RebuildFailed {
        return faulted_children_remover(nexus_spec, &child.uri, context).await;
    }

    if let Ok(replica) = get_child_replica(nexus_spec.as_ref(), child, context).await {
        info!(%child.uri, "Child's Replica is Online within the partial rebuild window");
        if online_nexus_child(nexus_spec, child, replica, context).await.is_err() {
            // we can also not do this and simply wait till the timeout..
            faulted_children_remover(nexus_spec, &child.uri, context).await?
        }
    }

    // Partial rebuild time window expired, we have to rebuild anew.
    if partial_rebuild_elapsed(child.faulted_at, &child.uri) {
        info!(%child.uri, "Child's Replica failed to become online within the partial rebuild window");
        return faulted_children_remover(nexus_spec, &child.uri, context).await;
    }
    
    Ok(())
}


```

Reply:
I have combined checking time_elapsed() and wait_time.is_zero() together. @dsharma-dc asked if i can combine it.

Reply:
Shouldnt have different if statements in this. code should enter anyone of all checks. Just tested, even if remove child (1st condition) we attempt get_child_replica. I am checking time elapsed before anything else.